tempesta_fw/http_sess.c

```diff
@@ -957,8 +960,37 @@ tfw_http_sess_obtain(TfwHttpReq *req)
 			continue;
 		}
 
-		if (!memcmp_fast(sv.hmac, sess->hmac, sizeof(sess->hmac)))
+		if (!memcmp_fast(sv.hmac, sess->hmac, sizeof(sess->hmac))) {
+			read_lock(&sess->lock);
+			if (!sess->vhost) {
+				read_unlock(&sess->lock);
+				goto found;
+			}
+
+			if (unlikely(test_bit(TFW_VHOST_B_INVALID,
+					      &sess->vhost->flags)))
+			{
+				/*
+				 * The session holds the last reference to the
+				 * vhost. It's not a part of the active
+				 * configuration anymore and therefore is not
+				 * useful to us at all. We can only release it
+				 * to free associated resources.
+				 */
+				TfwVhost *cached_vhost = sess->vhost;
+				read_unlock(&sess->lock);
+				write_lock(&sess->lock);
+				sess->vhost = NULL;
+				write_unlock(&sess->lock);
+				tfw_vhost_put(cached_vhost);
+				goto found;
+			}
+
+			tfw_vhost_get(sess->vhost);
+			req->vhost = sess->vhost;
+			read_unlock(&sess->lock);
 			goto found;
```

Comment:
Something seemed strange to me here, I was thinking about possible configuration updates like scheduler changes. vhost name changes and other changes that can spoil moving session from old vhost instance to a new one. I took me some time to recheck this, but everything looks good here.